pilot/pkg/model/jwks_resolver.go

```diff
@@ -336,25 +338,36 @@ func (r *JwksResolver) getRemoteContentWithRetry(uri string, retry int) ([]byte,
 
 func (r *JwksResolver) refresher() {
 	// Wake up once in a while and refresh stale items.
+	r.mu.Lock()
```

Comment:
I really think we should not use the lock if its only purpose is to make the unit test easier, the code gets much more complex and error prone with very small benefits, could you change to something like the below:

```golang
var (
	// JwtPubKeyRefreshIntervalOnFailureResetThreshold is the threshold to reset the refresh interval on failure.
	JwtPubKeyRefreshIntervalOnFailureResetThreshold = 60 * time.Minute
)

func (r *JwksResolver) refresher() {
	// Wake up once in a while and refresh stale items.
	r.refreshTicker = time.NewTicker(r.refreshInterval)
	lastHasError := false
	for {
		select {
		case <-r.refreshTicker.C:
			currentHasError := r.refresh()
			if currentHasError {
				if lastHasError {
					// update to exponential backoff if last time also failed.
					r.refreshInterval *= 2
					if r.refreshInterval > JwtPubKeyRefreshIntervalOnFailureResetThreshold {
						r.refreshInterval = r.refreshIntervalOnFailure
					}
				} else {
					// change to the refreshIntervalOnFailure if failed for the first time.
					r.refreshInterval = r.refreshIntervalOnFailure
				}
			} else {
				// reset the refresh interval if success.
				r.refreshInterval = r.refreshDefaultInterval
			}
			lastHasError = currentHasError
			r.refreshTicker.Stop()
			r.refreshTicker = time.NewTicker(r.refreshInterval)
		case <-closeChan:
			r.refreshTicker.Stop()
			return
		}
	}
}
```

And inside the refresh(), remove the code that updates the `r.refreshInterval`, the refresh() should only care about the single refresh logic, so let's move the backoff code to the outer level.

And in the test, you could do something like the below:

```golang
func TestJwtRefreshIntervalExponentialBackoff(t *testing.T) {
	defaultRefreshInterval := 50*time.Millisecond
	refreshIntervalOnFail := 2*time.Millisecond
	r := newJwksResolver(JwtPubKeyEvictionDuration, defaultRefreshInterval, refreshIntervalOnFail, 1*time.Millisecond)

	ms := startMockServer(t)
	defer ms.Stop()

	// Configures the mock server to return error after the first request.
	ms.ReturnErrorAfterFirstNumHits = 1

	mockCertURL := ms.URL + "/oauth2/v3/certs"
	_, err := r.GetPublicKey("", mockCertURL)
	if err != nil {
		t.Fatalf("GetPublicKey(%q, %+v) fails: expected no error, got (%v)", "", mockCertURL, err)
	}

	time.Sleep(100*time.Millisecond)
	r.Close()

	powerOfTwo := func(n int64) bool {
		return n != 0 && (n & (n - 1) == 0)
	}
	r.refreshInterval.Milliseconds()
	if r.refreshInterval == refreshIntervalOnFail || !powerOfTwo(r.refreshInterval.Milliseconds() / time.Millisecond.Milliseconds()) {
		t.Errorf("refreshInterval not updated with exponential backoff, got %v", r.refreshInterval)
	}
}
```

And you can add another `TestJwtRefreshIntervalRecoverFromFail()`, and set `ReturnErrorForFirstNumHits=3`, sleep a little bit, stop the refresher and then check the `r.refreshInterval == defaultRefreshInterval`.

Reply:
Thanks @yangminzhu. This makes totally sense to me.

Reply:
@yangminzhu I failed to figure out if `r.GetPublicKey` is getting called periodically. Since this is the only way the cache `r.keyEntries` is getting populated there is currently an issue for which I've added the test `TestJwtRefreshIntervalRecoverFromInitialFailOnFirstHit`. It illustrates the current issue.

I could fix this in that I extend the jwtPubKeyEntry to have a isValid field but I'm curious if that is needed. I think that depends on whether `r.GetPublicKey` is periodically called or not.
```golang
type jwtPubKeyEntry struct {
	pubKey string

	// The last success refreshed time of the pubKey.
	lastRefreshedTime time.Time

	// Cached item's last used time, which is set in GetPublicKey.
	lastUsedTime time.Time

        // isValid indicates whether the public key is valid to use.
        isValid bool
}
```

Using it in `r.GetPublicKey` like this:
```golang
	if jwksURI == "" {
		// Fetch the jwks URI if it is not hardcoded on config.
		var err error
		jwksURI, err = r.resolveJwksURIUsingOpenID(issuer)
		if err != nil {
			log.Errorf("Failed to jwks URI from %q: %v", issuer, err)
	                r.keyEntries.Store(key, jwtPubKeyEntry{
		                lastRefreshedTime: now,
		                lastUsedTime:      now,
                                isValid: false,
	                }
			return "", err
		}
	}
```

This way the key would also get refreshed in `r.refresh` and would trigger it into failure mode to potentially recover more quickly. But as mentioned above I'm not 100% sure if this is needed.

I tried to trace back the pilot xds/lds config generation to see if `r.GetPublicKey` is getting invoked periodically. Since there is key deletion in `r.refresh` like below I guess (hope) that means there must be some sort of period config generation which triggers `r.GetPublicKey` even if there are no external CRD config changes otherwise with the code so far in this PR I fear that we could end up with the exact same issue for which I've opened this PR to fix it ;) Is that the case?
```golang

		// Remove cached item for either of the following 2 situations
		// 1) it hasn't been used for a while
		// 2) it hasn't been refreshed successfully for a while
		// This makes sure 2 things, we don't grow the cache infinitely and also we don't reuse a cached public key
		// with no success refresh for too much time.
		if now.Sub(e.lastUsedTime) >= r.evictionDuration || now.Sub(e.lastRefreshedTime) >= r.evictionDuration {
			log.Infof("Removed cached JWT public key (lastRefreshed: %s, lastUsed: %s) from %q",
				e.lastRefreshedTime, e.lastUsedTime, k.issuer)
			r.keyEntries.Delete(k)
			return true
		}
```

Reply:
`GetPublicKey()` is called when a RequestAuthentication is created or updated, if the jwksUri is fetched successfully without error, the jwksUri will be stored to the `keyEntries` and the refresh job will periodically update the jwksUri in the `keyEntries`.

I think you don't need to worry about that `GetPublicKey()` is not called periodically, but the real problem is when `GetPublickKey()` failed, it won't store the jwksUri to the `keyEntries` which means it won't be refreshed by the periodic job. So I think you could fix this by changing `GetPublicKey()` to always store the jwksUri to the `kenEntries` even if it fails (you can just use an empty pubKey, we called `createFakeJwks()` in policy_applier.go to handle the error case)

That's a very good finding and I think it's not covered in the initial PR, and should be tested with the `TestJwtRefreshIntervalRecoverFromInitialFailOnFirstHit()`, thank you!

Reply:
Thanks a lot I've addressed that.

Reply:
@yangminzhu I've fixed all the outstanding issues and rebased to latest upstream which still looks ok though I don't understand why this PR now shows the commits of pull requests which were approved and are already in master. Anyway I can also open another PR as there were some merge conflicts with #29704 which I did fix.

Reply:
Just for the record also all the merge checks are now satisfied.

Reply:
@yangminzhu I've created another branch without the merge conflicts. Please let me know if I should open a PR for this: https://github.com/istio/istio/compare/master...domcyrus:fix-jwks-uri-on-startup3
and close this one. Thanks.